### ISSUE TITLE
Captagent is utilizing 100% CPU  when sip-parse = true

### ISSUE THREAD

**Issue by marrold** [sipcapture/captagent#5]
Captagent is utilizing 100% CPU  when sip-parse = true

This doesn't happen as soon as Captagent is started. It can take 2-5 minutes before it pegs at 100% CPU utilization. 

Server Details-
Intel(R) Xeon(R) CPU E5504  @ 2.00GHz (Quad Core)
4 GB RAM

Traffic Details-
rx: 119.94 Mbit/s 77492 p/s 
SIP MPS Approx 160

capagent.xml -

<?xml version="1.0"?>

<document type="captagent/xml">

        <configuration name="core.conf" description="CORE Settings">
          <settings>
            <param name="debug" value="-3"/>
            <param name="daemon" value="true"/>
            <param name="syslog" value="false"/>
            <param name="pid_file" value="/var/run/captagent1.pid"/>
            <param name="path" value="/usr/local/lib/captagent/modules"/>
          </settings>
        </configuration>

        <configuration name="modules.conf" description="Modules">
          <modules>
                <load module="core_hep"/>
                <load module="proto_uni"/>
          </modules>
        </configuration>

        

        <configuration name="core_hep.conf" description="HEP Socket">
          <settings>
            <param name="version" value="3"/>
            <param name="capture-host" value="127.0.0.1"/>
            <param name="capture-port" value="9061"/>
            <param name="capture-proto" value="udp"/>
            <param name="capture-id" value="1"/>
            <param name="capture-password" value="myHep"/>
            <param name="payload-compression" value="false" />
          </settings>
        </configuration>

        

        <configuration name="proto_uni.conf" description="UNI Proto Basic capture">
          <settings>
            <param name="portrange" value="5060-5080"/>
            
            
            <param name="dev" value="em2"/>
            <param name="promisc" value="true"/>
            
            
            <param name="proto-type"  value="sip"/>
            <param name="sip-parse"  value="true"/>
            <param name="rtcp-tracking"  value="false"/>
            <param name="reasm"  value="false"/>
            <param name="tcpdefrag"  value="true"/>
            <param name="debug" value ="false"/>
            <param name="buildin-reasm-filter" value="false"/>
            
            
                
                
          </settings>
        </configuration>
</document>

Thanks in advance.

**Comment by adubovikov** [comment]
I have checked, but unfortunately couldn't reproduce the issue. For me it looks like bad SIP parsing, Can you please crate a pcap dump with all SIP messages from start until high CPU usage ? After I will check again. Thank you.

**Comment by marrold** [comment]
Hi Alexandr,

That's plausible. I'm afraid sending a full PCAP isn't really possible as
it contains sensitive data as I'm sure you can imagine. Would increasing
the verbosity on the captagent logs help?

I could try capturing an offending packet this evening when there is less
traffic, to make it more straight forward to redact any numbers / IP's etc

Thanks

**Comment by adubovikov** [comment]
please check last git. 
i did on workaround, please check syslog/log file if you will see something like this:
"TOO MANY LOOPS [10]"

thank you.

**Comment by marrold** [comment]
Thanks, I will test and report back.

**Comment by marrold** [comment]
100% CPU reoccurred today. Please see the logs- http://pastebin.com/raw.php?i=wWF0TdjM

Also, captagent died with a segfault, possibly due to 100% CPU usage -

Jan 26 10:38:09 CLL-Tracing kernel: [926734.544000] captagent[6523]: segfault at 7fb51c12230f ip 00007fb4248a9b8e sp 00007fb4239c05c8 error 4 in libc-2.19.so[7fb424811000+1bb000]

I don't have the packet capture for this occurrence, I can grab one next time if required?

Thanks

**Comment by adubovikov** [comment]
no, it's ok for now. I see that len is bigger than message itself. I have added more debug, can you please pull git again and check one more time ?

**Comment by marrold** [comment]
Pulled and running, Thanks

**Comment by marrold** [comment]
Please see a small selection of the latest log showing the common messages. As you believe it's a len issue I've been careful to use the same amount of characters when redacting numbers / IP's

http://pastebin.com/raw.php?i=GnraNnMW

Thanks

**Comment by marrold** [comment]
Hi, is there any update on this? I noticed it's no longer utilising 100% CPU due to the added 'break', but curious if a fix is on its way.

Thanks

**Comment by adubovikov** [comment]
Yes, we will update tomorrow. Sorry but this week was very busy.

**Comment by marrold** [comment]
No need to apologies, appreciated as always.

On Sun, Feb 8, 2015 at 9:15 PM, Alexandr Dubovikov <notifications@github.com
> wrote:

> Yes, we will update tomorrow. Sorry but this week was very busy.
>
>

**Comment by adubovikov** [comment]
so. please take the last git. Also I see that this TCP messages are broken. Probably it was bad message len, but it's hard to say.

anyway I am waiting on your feedback. Thank you!

**Comment by marrold** [comment]
Thanks, I will test tomorrow. Is there anyway to disregard broken TCP and
parse anyway?

**Comment by adubovikov** [comment]
hard to say. If only some lines of SDP are missed, it's ok, but if RURI or several important headers are missed - this is complete different story. Anyway, kamailio will drop it :-)

**Comment by marrold** [comment]
What I am trying to understand is, why wasn't the below example (from the
previous pastebin) parsed? It looks RFC compliant. Was this a TCP LEN vs
Actual Data LEN issue? Will the latest code mean such SIP messages are
parsed?

Thanks

Jan 27 18:25:44 CLL-Tracing captagent[27177]: [ERR] proto_uni.c:328 TOO
MANY LOOP LEN [1386] vs NEWLEN: [0] vs SKIP: [1383] vs PARSED: [0]
Jan 27 18:25:44 CLL-Tracing captagent[27177]: [ERR] proto_uni.c:329 PACKET

INVITE sip:111111111111@172.16.1.23:5060;transport=tcp SIP/2.0
Via: SIP/2.0/TCP 11.111.1.111:5060
;branch=z9hG4bK-524287-1---d5a2e626e94acc0d;rport
Via: SIP/2.0/UDP 11.111.1.111:5061
;branch=z9hG4bK-6x3eokgxtich7y4w;rport=5061
Max-Forwards: 69
Record-Route: <sip:11.111.1.111:5060;transport=tcp;lr;drr>
Record-Route: <sip:11.111.1.111:5060;lr;transport=UDP;drr>
Contact: "Anonymous"<sip:11.111.1.111:5061>
To: <sip:111111111111@172.16.1.23>
From: +111111111111 <sip:+111111111111@11.111.1.111>;tag=nkgcuhqzbwycb74v.o
Call-ID: 456b4fac-333cbed0-7f5fb838-a72a@10.16.231.133
CSeq: 781 INVITE
Expires: 300
Content-Disposition: session
Content-Type: application/sdp
User-Agent: Sippy
P-Asserted-Identity: +111111111111 <sip:+111111111111@11.111.1.111
;user=phone>
h323-conf-id: 235399844-542757103-3421031868-2401107647
Portasip-3264-action: offer 1
cisco-GUID: 235399844-542757103-3421031868-2401107647
Content-Length: 445

v=0
o=Sippy 2712874956379544410 0 IN IP4 11.111.1.111
s=Cisco SDP 0
t=0 0
m=audio 64644 RTP/AVP 8 18 0 101
c=IN IP4 11.111.1.111
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-15
a=sqn:0
a=cdsc: 1 audio RTP/AVP 8 18 0 101
a=cdsc: 5 image udptl t38
a=cpar: a=T38FaxVersion:0
a=cpar: a=T38FaxRateManagement:transferredTCF
a=cpar: a=T38FaxMaxDatagram:160
a=cpar: a=T38FaxUdpEC:t38UDPRedundancy
a=X-sqn:0
a=X-cap: 1 image udptl t38


On Mon, Feb 9, 2015 at 9:05 PM, Alexandr Dubovikov <notifications@github.com

**Comment by adubovikov** [comment]
the message len is 1386, but we have only 1383, somethere 3 characters are gone.
anyway, with the new git, this message will be parsed.

**Comment by adubovikov** [comment]
any feedback ? Can I close the issue ?

**Comment by marrold** [comment]
I spotted one entry in the logs regarding the loop, which is much less than
before. I was going to leave it 24 hours before providing feed back.

There still seems to be some issues with TCP packets, which may or may not
be related.

I don't suppose someone has created a HEP dissector for Wireshark?

Thanks

**Comment by adubovikov** [comment]
yeah, we planned do it long time ago, but unfortunately no time for this. If you can do it, we will be very appreciated!!!

**Comment by marrold** [comment]
I plan to give it a go but can't guarantee any success!

In the mean time, should I raise a new issue for the missing TCP packets
once I've gathered some examples? I'm happy to wait as you've been busy.

**Comment by adubovikov** [comment]
sure, contact me any time. Currently we have some requests,  but we always find a bit time to help our users :-)

**Comment by marrold** [comment]
Please find the latest log here- https://gist.github.com/marrold/00e7e647ae0ece7e46f8

Looks like only 3 packets got caught in the loop out of thousands, so this is much better.

Thanks

**Comment by adubovikov** [comment]
ok. looks like it was fixed. Please check the last git.

thank you very much!

**Comment by marrold** [comment]
Pulled and running. Thanks for your assistance.

**Comment by adubovikov** [comment]
you are welcome. thank for bug reporting.
